File: deploy/packaging/macos/merge-launchd-plist.py
#!/usr/bin/env python3
"""Preserve safe existing LaunchDaemon options during a bundle transition."""

from __future__ import annotations

import argparse
import os
import plistlib
import tempfile
from pathlib import Path


_FIXED_WITH_VALUE = {"--plugin-dir", "--data-dir", "--log-file", "--log-max-size", "--log-max-files"}
_SIGNING_OPTIONS = {"--plugin-trust-bundle", "--plugin-require-signature"}
_PRESERVED_ENVIRONMENT = {
    "YUZU_SERVER", "YUZU_CA_CERT", "YUZU_TLS_SYSTEM_ROOTS", "YUZU_CLIENT_CERT", "YUZU_CLIENT_KEY",
    "YUZU_CERT_STORE", "YUZU_CERT_SUBJECT", "YUZU_CERT_THUMBPRINT", "YUZU_PLUGIN_ALLOWLIST",
    "YUZU_PLUGIN_TRUST_BUNDLE", "YUZU_PLUGIN_REQUIRE_SIGNATURE",
}


def merge(previous: Path, destination: Path, *, force_no_auto_update: bool = False) -> None:
    with previous.open("rb") as source:
        prior = plistlib.load(source)
    with destination.open("rb") as source:
        generated = plistlib.load(source)
    old_args = prior.get("ProgramArguments")
    new_args = generated.get("ProgramArguments")
    if not isinstance(old_args, list) or not all(isinstance(arg, str) for arg in old_args):
        raise ValueError("previous plist has invalid ProgramArguments")
    if not isinstance(new_args, list) or not new_args or not all(isinstance(arg, str) for arg in new_args):
        raise ValueError("generated plist has invalid ProgramArguments")

    generated_signing = "--plugin-trust-bundle" in new_args
    preserved: list[str] = []
    index = 1  # executable path is always replaced by the generated bundle path.
    while index < len(old_args):
        argument = old_args[index]
        name, separator, _value = argument.partition("=")
        if name in _FIXED_WITH_VALUE:
            index += 2 if not separator else 1
            continue
        if name == "--no-auto-update" and force_no_auto_update:
            index += 1
            continue
        if generated_signing and name in _SIGNING_OPTIONS:
            index += 2 if name == "--plugin-trust-bundle" and not separator else 1
            continue
        preserved.append(argument)
        index += 1

    generated["ProgramArguments"] = [*new_args, *preserved]
    old_environment = prior.get("EnvironmentVariables", {})
    if not isinstance(old_environment, dict) or not all(isinstance(key, str) and isinstance(value, str)
                                                         for key, value in old_environment.items()):
        raise ValueError("previous plist has invalid EnvironmentVariables")
    retained_environment = {
        key: value for key, value in old_environment.items() if key in _PRESERVED_ENVIRONMENT
    }
    if generated_signing:
        retained_environment.pop("YUZU_PLUGIN_TRUST_BUNDLE", None)
        retained_environment.pop("YUZU_PLUGIN_REQUIRE_SIGNATURE", None)
    if retained_environment:
        generated["EnvironmentVariables"] = retained_environment
    else:
        generated.pop("EnvironmentVariables", None)
    fd, temporary = tempfile.mkstemp(prefix="yuzu-plist-", dir=destination.parent)
    try:
        with os.fdopen(fd, "wb") as output:
            plistlib.dump(generated, output, sort_keys=False)
        os.replace(temporary, destination)
    except Exception:
        os.unlink(temporary)
        raise


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--previous", required=True, type=Path)
    parser.add_argument("--destination", required=True, type=Path)
    parser.add_argument("--force-no-auto-update", action="store_true")
    args = parser.parse_args()
    merge(args.previous, args.destination, force_no_auto_update=args.force_no_auto_update)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: deploy/packaging/macos/merge-launchd-plist.js
// Run with /usr/bin/osascript -l JavaScript. Uses only macOS Foundation;
// installation does not require Python, Xcode, a shell eval, or GUI automation.
ObjC.import('Foundation');

function readPlist(path) {
    const data = $.NSData.dataWithContentsOfFile(path);
    if (!data) throw new Error('Cannot read plist');
    const error = Ref();
    const value = $.NSPropertyListSerialization.propertyListWithDataOptionsFormatError(data, 0, null, error);
    if (!value) throw new Error('Cannot decode plist');
    return ObjC.deepUnwrap(value);
}

function merge(prior, generated, forceNoAutoUpdate) {
    const oldArgs = prior.ProgramArguments;
    const newArgs = generated.ProgramArguments;
    function stringArray(value) {
        return Array.isArray(value) && value.length > 0 && value.every(x => typeof x === 'string');
    }
    if (!stringArray(oldArgs) || !stringArray(newArgs)) throw new Error('Invalid ProgramArguments');
    const fixed = ['--plugin-dir', '--data-dir', '--log-file', '--log-max-size', '--log-max-files'];
    const signing = ['--plugin-trust-bundle', '--plugin-require-signature'];
    const generatedSigning = newArgs.indexOf('--plugin-trust-bundle') >= 0;
    const preserved = [];
    for (let index = 1; index < oldArgs.length;) {
        const argument = oldArgs[index];
        const name = argument.split('=', 1)[0];
        const inlineValue = argument.indexOf('=') >= 0;
        if (fixed.indexOf(name) >= 0) {
            index += inlineValue ? 1 : 2;
        } else if (name === '--no-auto-update' && forceNoAutoUpdate) {
            index += 1;
        } else if (generatedSigning && signing.indexOf(name) >= 0) {
            index += name === '--plugin-trust-bundle' && !inlineValue ? 2 : 1;
        } else {
            preserved.push(argument);
            index += 1;
        }
    }
    generated.ProgramArguments = newArgs.concat(preserved);
    const allowed = ['YUZU_SERVER', 'YUZU_CA_CERT', 'YUZU_TLS_SYSTEM_ROOTS', 'YUZU_CLIENT_CERT',
        'YUZU_CLIENT_KEY', 'YUZU_CERT_STORE', 'YUZU_CERT_SUBJECT', 'YUZU_CERT_THUMBPRINT',
        'YUZU_PLUGIN_ALLOWLIST', 'YUZU_PLUGIN_TRUST_BUNDLE', 'YUZU_PLUGIN_REQUIRE_SIGNATURE'];
    const oldEnvironment = prior.EnvironmentVariables === undefined ? {} : prior.EnvironmentVariables;
    if (!oldEnvironment || Array.isArray(oldEnvironment) || typeof oldEnvironment !== 'object' ||
        !Object.keys(oldEnvironment).every(key => typeof oldEnvironment[key] === 'string')) {
        throw new Error('Invalid EnvironmentVariables');
    }
    const retained = {};
    allowed.forEach(key => {
        if (Object.prototype.hasOwnProperty.call(oldEnvironment, key) &&
            !(generatedSigning && (key === 'YUZU_PLUGIN_TRUST_BUNDLE' || key === 'YUZU_PLUGIN_REQUIRE_SIGNATURE'))) {
            retained[key] = oldEnvironment[key];
        }
    });
    if (Object.keys(retained).length) generated.EnvironmentVariables = retained;
    else delete generated.EnvironmentVariables;
    return generated;
}

function run(argv) {
    let previous, destination, force = false;
    for (let index = 0; index < argv.length; index++) {
        if (argv[index] === '--previous') previous = argv[++index];
        else if (argv[index] === '--destination') destination = argv[++index];
        else if (argv[index] === '--force-no-auto-update') force = true;
        else throw new Error('Unknown argument');
    }
    if (!previous || !destination) throw new Error('Previous and destination plist paths required');
    const result = merge(readPlist(previous), readPlist(destination), force);
    const error = Ref();
    const data = $.NSPropertyListSerialization.dataWithPropertyListFormatOptionsError(
        $(result), $.NSPropertyListXMLFormat_v1_0, 0, error);
    if (!data || !data.writeToFileAtomically(destination, true)) throw new Error('Cannot write merged plist');
}
